executor/src/witgen/affine_expression.rs

```diff
@@ -444,28 +421,110 @@ where
-            self.clean = rhs.clean;
-        } else if let [(lk, lv)] = self.coefficients[..] {
-            self.clean = !rhs.coefficients.iter().any(|(k, _)| k == &lk);
-            self.coefficients = rhs.coefficients;
-            self.coefficients.push((lk, lv));
-        } else if let [(rk, rv)] = rhs.coefficients[..] {
-            self.clean = !self.coefficients.iter().any(|(k, _)| k == &rk);
-            self.coefficients.push((rk, rv));
-        } else {
-            self.coefficients.append(&mut rhs.coefficients);
-            self.clean = false;
+    fn add(self, rhs: Self) -> Self::Output {
+        use std::cmp::Ordering;
+        use AffineExpression::*;
+        match (self, rhs) {
+            // constant + constant
+            (Constant(o1), Constant(o2)) => Constant(o1 + o2),
+            // constant + one
+            (One(coeff, o1), Constant(o2)) | (Constant(o2), One(coeff, o1)) => One(coeff, o1 + o2),
+            // constant + many
+            (
+                Many {
+                    coefficients,
+                    offset,
+                },
+                Constant(o),
+            )
+            | (
+                Constant(o),
+                Many {
+                    coefficients,
+                    offset,
+                },
+            ) => Many {
+                coefficients,
+                offset: offset + o,
+            },
+            // one + one
+            (One((v1, c1), o1), One((v2, c2), o2)) => match v1.cmp(&v2) {
+                Ordering::Equal => {
+                    let c = c1 + c2;
+                    if c.is_zero() {
+                        Constant(o1 + o2)
+                    } else {
+                        One((v1, c), o1 + o2)
+                    }
+                }
+                Ordering::Less => Many {
+                    coefficients: vec![(v1, c1), (v2, c2)],
+                    offset: o1 + o2,
+                },
+                Ordering::Greater => Many {
+                    coefficients: vec![(v2, c2), (v1, c1)],
+                    offset: o1 + o2,
+                },
+            },
+            // one + many
+            (
+                One((v, c), o),
+                Many {
+                    mut coefficients,
+                    offset,
+                },
+            )
+            | (
+                Many {
+                    mut coefficients,
+                    offset,
+                },
+                One((v, c), o),
+            ) => {
+                debug_assert!(Self::is_proper_coefficients(&coefficients));
+                match coefficients.binary_search_by_key(&&v, |(k, _)| k) {
+                    Ok(idx) => coefficients[idx].1 += c,
+                    Err(idx) => coefficients.insert(idx, (v, c)),
+                }
+                coefficients.retain(|(_, c)| !c.is_zero());
+                if coefficients.is_empty() {
+                    Constant(offset + o)
+                } else {
```

Comment:
So this is the case where you could reduce to One but don't right?

Reply:
yes, there are two places where `coefficients.is_empty()` where we could also test for `len() == 1`